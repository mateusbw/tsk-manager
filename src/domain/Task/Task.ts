import { v4 as uuidv4 } from "uuid";

export enum Status {
  ToDo = "toDo",
  Closed = "closed",
}

export type Task = {
  id: string;
  title: string;
  description: string;
  status: Status;
  createdAt: number;
};

export const newTask = (title: string, description: string): Task => ({
  id: uuidv4(),
  title,
  description,
  status: Status.ToDo,
  createdAt: Date.now(),
});

export const closeTask = (task: Task): Task => ({
  ...task,
  status: Status.Closed,
});

export const openTask = (task: Task): Task => ({
  ...task,
  status: Status.ToDo,
});

export const isClosed = (task: Task): boolean => task.status === Status.Closed;
