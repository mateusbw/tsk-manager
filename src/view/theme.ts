const colors = {
  gray: {
    750: "#414052",
    850: "#272733",
    900: "#21212B",
    950: "#181820",
  },
  purple: "#AC6DDE",
  red: "#f93c3c",
  white: "#F5F5F6",
};

const spaces = {
  xs: ".25rem",
  sm: ".5rem",
  md: "1rem",
  lg: "1.5rem",
  xg: "3rem",
};

const fontSizes = {
  sm: "16px",
  md: "26px",
  lg: "42px",
};

const fontWeights = {
  thin: "300",
  medium: "400",
  semiBold: "600",
};

const theme = { colors, spaces, fontSizes, fontWeights };

export default theme;
