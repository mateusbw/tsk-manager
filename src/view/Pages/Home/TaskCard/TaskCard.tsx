import styled, { keyframes } from "styled-components";
import Flex from "../../../Components/Layout/Flex";
import Text from "../../../Components/Text/Text";
import CloseTaskButton from "../../../Components/CloseTaskButton/CloseTaskButton";
import Card from "../../../Components/Card/Card";
import CardHeader from "../../../Components/Card/CardHeader/CardHeader";
import CardBody from "../../../Components/Card/CardBody/CardBody";
import CardFooter from "../../../Components/Card/CardFooter/CardFooter";
import Button from "../../../Components/Button/Button";
import { ReactComponent as DotsIcon } from "../../../../assets/icons/dots.svg";

import theme from "../../../theme";
import { isClosed as isTaskClosed, Task } from "../../../../domain/Task/Task";
import timeUtil from "../../../../utils/timeUtil";

const onEnterRight = keyframes`
    from{opacity: 0; transform: translateX(50%)}
    to{opacity: 1; transform: translateX(0)}`;

const onEnterLeft = keyframes`
    from{opacity: 0; transform: translateX(-50%)}
    to{opacity: 1; transform: translateX(0)}
    `;

const BaseTaskCard = styled(Card)<{ isClosed: boolean }>`
  animation: ${({ isClosed }) => (isClosed ? onEnterLeft : onEnterRight)} 0.5s
    forwards;
`;

type Props = {
  task: Task;
  markAsClosed: () => void;
  markAsToDo: () => void;
};

const TaskCard = ({ task, markAsClosed, markAsToDo }: Props) => {
  return (
    <BaseTaskCard
      mb="20px"
      isClosed={isTaskClosed(task)}
      data-testid="task-card"
    >
      <CardHeader display="flex" justifyContent="space-between">
        <Flex alignItems="center">
          <CloseTaskButton
            isClosed={isTaskClosed(task)}
            onCloseCallback={markAsClosed}
            onTodoCallback={markAsToDo}
          />
          <Text as="h2" variant="small" fontWeight={theme.fontWeights.semiBold}>
            {task.title}
          </Text>
        </Flex>
        <Button variant="icon" p="0">
          <DotsIcon />
        </Button>
      </CardHeader>
      <CardBody>
        <Text as="p" variant="small">
          {task.description}
        </Text>
      </CardBody>
      <CardFooter textAlign="center">
        <Text as="p" variant="small" fontWeight={theme.fontWeights.thin}>
          Created {timeUtil.formatDate(task.createdAt, "ff")}
        </Text>
      </CardFooter>
    </BaseTaskCard>
  );
};

export default TaskCard;
