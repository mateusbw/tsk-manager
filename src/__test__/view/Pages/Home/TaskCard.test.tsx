import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { newTask } from "../../../../domain/Task/Task";
import TaskCard from "../../../../view/Pages/Home/TaskCard/TaskCard";
import theme from "../../../../view/theme";

describe("<TaskCard/>", () => {
  it("matches   snapshot", () => {
    Date.now = jest.fn(() => new Date("2019-04-07T10:20:30Z").getTime());
    const task = newTask("Title", "Description");
    const taskCardComponent = render(
      <ThemeProvider theme={theme}>
        <TaskCard
          task={task}
          markAsClosed={() => null}
          markAsToDo={() => null}
        />
      </ThemeProvider>
    );
    expect(taskCardComponent.container).toMatchSnapshot();
  });
});
