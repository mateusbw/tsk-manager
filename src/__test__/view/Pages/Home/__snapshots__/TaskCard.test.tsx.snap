// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskCard/> matches   snapshot 1`] = `
<div>
  <div
    class="sc-jrsJWt bpNRWa sc-dIsUp ktQXLX"
  >
    <div
      class="sc-kEqXSa hPuJQS"
      display="flex"
    >
      <div
        class="sc-bdnxRM dQuBew"
      >
        <button
          class="sc-fujyAs jEaHlk sc-pNWdM gmOIvx"
        />
        <h2
          class="sc-dlnjwi cdEQQn"
          font-weight="600"
        >
          Title
        </h2>
      </div>
      <button
        class="sc-fujyAs bOHUEs"
      >
        <svg>
          dots.svg
        </svg>
      </button>
    </div>
    <div
      class="sc-iqAclL imHOEE"
    >
      <p
        class="sc-iCoGMd fCDwTH"
      >
        Description
      </p>
    </div>
    <div
      class="sc-crzoAE iUJPTx"
    >
      <p
        class="sc-iCoGMd fIWBnv"
        font-weight="300"
      >
        Created 
        Apr 7, 2019, 7:20 AM
      </p>
    </div>
  </div>
</div>
`;
